File: problems/059_regiuni_2/UnionFind.pas
type 
  UnionFind = class
  private n : integer;
  private id : array of integer;
  private sz : array of integer;
  private count : integer;
  public constructor Create(num : integer);
  var i : integer;
  begin
    n := num;
    count := num;
    setLength(id, n);
    setLength(sz, n);
    for i := 0 to n - 1 do
    begin
      id[i] := i;
      sz[i] := 1;
    end;
  end;
  
  function getCount : integer;
  begin
    Result := count;
  end;
  
  function find(i : integer) : integer;
  begin
    while (i <> id[i]) do
    begin
      i := id[i];
    end;
    Result := i;
  end;
  
  procedure union(p, q : integer);
  var i, j : integer;
  begin
    i := find(p);
    j := find(q);
    {writeln('i = ', i, ' j = ', j);}
    if (i = j) then
    begin
      exit;
    end;
    if (sz[i] < sz[j]) then
    begin
      id[i] := j;
      sz[j] := sz[j] + sz[i];
    end
    else 
    begin
      id[j] := i;
      sz[i] := sz[i] + sz[j];
    end;
    count := count - 1;
  end;
end;
 
var
  num : integer;
  uf : UnionFind;
  i : integer;
begin
  num := 5;
  uf := new UnionFind(num);
  for i := 0 to num - 1 do
  begin
    writeln('uf[', i, '] = ', uf.id[i]);
  end;
end.
  

File: computer_science/059_regiuni_2/UnionFind.java
import edu.princeton.cs.algs4.StdOut;

public class UnionFind {
    private final int[] id;
    private final int[] sz;
    private int count;

    public UnionFind(int n) {
        this.count = n;
        id = new int[n];
        for (int i = 0; i < n; i++) {
            id[i] = i;
        }

        sz = new int[n];
        for (int i = 0; i < n; i++) {
            sz[i] = 1;
        }
    }

    public int count() {
        return count;
    }

    public int find(int i) {
        while (i != id[i]) {
            i = id[i];
        }
        return i;
    }

    public void union(int p, int q) {
        int i = find(p);
        int j = find(q);
        if (i == j) {
            return;
        }
        if (sz[i] < sz[j]) {
            id[i] = j;
            sz[j] += sz[i];
        }
        else {
            id[j] = i;
            sz[i] += sz[j];
        }
        count--;
        }

    public static void main(String[] args) {
        int n = Integer.parseInt(args[0]);
        UnionFind UF = new UnionFind(n);
        UF.union(1, 2);
    }
}
